src/pip/_internal/operations/prepare.py

```diff
@@ -517,41 +521,92 @@ def prepare_linked_requirement(
                 # The file is not available, attempt to fetch only metadata
                 metadata_dist = self._fetch_metadata_only(req)
                 if metadata_dist is not None:
-                    req.needs_more_preparation = True
+                    # These reqs now have the dependency information from the downloaded
+                    # metadata, without having downloaded the actual dist at all.
+                    req.cache_virtual_metadata_only_dist(metadata_dist)
                     return metadata_dist
 
             # None of the optimizations worked, fully prepare the requirement
             return self._prepare_linked_requirement(req, parallel_builds)
 
-    def prepare_linked_requirements_more(
-        self, reqs: Iterable[InstallRequirement], parallel_builds: bool = False
+    def _ensure_download_info(self, reqs: Iterable[InstallRequirement]) -> None:
+        """
+        `pip install --report` extracts the download info from each requirement for its
+        JSON output, so we need to make sure every requirement has this before finishing
+        the resolve. But .download_info will only be populated by the point this method
+        is called for requirements already found in the wheel cache, so we need to
+        synthesize it for uncached results. Luckily, a DirectUrl can be parsed directly
+        from a url without any other context. However, this also means the download info
+        will only contain a hash if the link itself declares the hash.
```

Comment:
This is the change I was calling out in another comment. I don't know if it can be considered a real breaking change but it's worth thinking about it.

Dry-run without actually downloading the wheels is definitely important. I'm wondering, though, if there are use case for a dry-run where the packages are downloaded (and cached) and hashes (possibly with different algorithms) computed and reported. For users wanting a "lock then install" workflow, this maybe desirable.

Reply:
I think you're absolutely right! In fact, the functionality you describe is probably exactly what I would need for use in pex e.g. pantsbuild/pex#2210 (because pex's pipelined installation process currently invokes `pip download`).

Let's say we add another flag `--ensure-downloaded-hashes[=sha256,sha512]`?

However, thinking about this use case also gave me another idea:
> For users wanting a "lock then install" workflow, this maybe desirable.

I would love for pip to support this more directly. I'm wondering whether you would accept a separate PR to add:
- `pip install --ensure-downloaded-hashes[=sha256,sha512]` (used with `--report` to ensure `.download_info` is populated)
- `pip install --from-report <file.json>` (can be used with `--require-hashes`)

I am not sure how most users are currently processing `--report` output, though. I'm also not sure if there's another lockfile format we would want to convert to instead of using the `--report` output as a lockfile directly? If so, we could maybe put it somewhere like `pip freeze --format=[requirements,lockfile] --from-report <file.json>`?

I really like the idea of explicitly supporting this "lock then install" workflow, and I would like to know your thoughts on whether pip could do that. Please also let me know if this workflow is already supported. cc @uranusjr @pradyunsg @pfmoore

Reply:
> I really like the idea of explicitly supporting this "lock then install" workflow, and I would like to know your thoughts on whether pip could do that. Please also let me know if this workflow is already supported.

I don't think it's something pip needs to support directly, but it's something we can provide components for. Lock-then-install workflows are available in tools like Poetry and PDM, and I don't think we need to compete with them in this area.

> `pip install --from-report <file.json>` (can be used with `--require-hashes`)

-1 on this. Processing the JSON to produce a requirements file (for now) or a standard lockfile (once such a format gets standardised) should, IMO, be the way forward here. We should focus on providing the low-level capabilities that higher-level workflow tools can be built from, not on pip becoming a workflow tool itself.

> `pip install --ensure-downloaded-hashes[=sha256,sha512]` (used with `--report` to ensure `.download_info` is populated)

I'm not particularly comfortable with this for the same reason - if pip *has* the hashes, it'll add them to the report. If it *doesn't* have them, then a tool wrapping pip can just as easily add them for itself by downloading the file and computing the hash. Again, if pip were managing the whole workflow, I might have a different view, but I don't see that as pip's role.

See https://github.com/pypa/pip/pull/12503 for more details on pip's role here.